llvm/labs/lab2/durandin_vladimir/vladimir_inline_pass.cpp

```diff
@@ -0,0 +1,165 @@
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include <string>
+
+using namespace llvm;
+
+namespace {
+Instruction *getNextNonTerminator(BasicBlock *BB, Instruction *I) {
+  for (auto &Inst : *BB) {
+    if (&Inst == I) {
+      for (auto &NextInst :
+           llvm::make_range(std::next(I->getIterator()), BB->end())) {
+        if (!NextInst.isTerminator()) {
+          return &NextInst;
+        }
+      }
+    }
+  }
+  return nullptr;
+}
+
+struct CustomInliningPass : public PassInfoMixin<CustomInliningPass> {
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &AM) {
+
+    Function *Callee{};
+    std::vector<CallInst *> callsToInline{};
+    for (auto &BB : F) {
+      for (auto &I : BB) {
+        if (auto *CI = dyn_cast<CallInst>(&I)) {
+          Callee = CI->getCalledFunction();
+          if (Callee && Callee->arg_size() == 0 &&
+              Callee->getReturnType()->isVoidTy()) {
+            callsToInline.push_back(CI);
+          }
+        }
+      }
+    }
+
+    // Process each call site
+    size_t counter_splited = 0;
+    size_t counter_inlined = 0;
+    for (CallInst *CI : callsToInline) {
+      BasicBlock *InsertBB = CI->getParent();
+      Instruction *InsertPt = getNextNonTerminator(InsertBB, CI);
+
+      // Create a map for block correspondence
+      DenseMap<BasicBlock *, BasicBlock *> BlockMap;
+
+      // Split the calling function at the call site (use ".splited." for
+      // clarity)
+      BasicBlock *SplitBB = InsertBB->splitBasicBlock(
+          InsertPt, InsertBB->getName() + ".splited." +
+                        std::to_string(counter_splited++));
+
+      // Create new blocks corresponding to empty function's blocks
+      BasicBlock *CalleeEntryBasicBlockPointer = nullptr;
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BasicBlock::Create(
+            F.getContext(),
+            CalleeBB.getName() + ".inlined." + std::to_string(counter_inlined),
+            &F);
+        if (Callee->getEntryBlock().getName() == CalleeBB.getName())
```

Comment:
```suggestion
        if (Callee->getEntryBlock() == CalleeBB)
```

Reply:
There is no `== operator` corresponding to these operands. And I’m comparing the names of the basic blocks, and not the pointers to them themselves, if that’s what you mean.

Reply:
Comparing pointers is more effective than strings.

Reply:
This, of course, is all understandable, but initially my logic was to get a pointer to the required base block. And since the functions are different, the basic blocks could be called the same.
In general, in order not to get confused for all of us, I simplified the logic just according to your advice and now there are no such moments :)